### PR TITLE
Refactored AEM Groovy Console to be able to run in Apache Sling. Remo…

### PR DESCRIPTION
**Pull request by royteeuwen**
…ved all Node usage and work based on Resource

### CODE REVIEW

**Comment by royteeuwen** [comment]
To Do: Create a basic console page in Sling to run the Groovy Console. Test in AEM

File: README.md [hunk 1]
```diff
@@ -17,26 +17,20 @@ scripts are included in the package for reference.
...
-* AEM author instance running on [http://localhost:4502](http://localhost:4502/)
-* [Maven](https://maven.apache.org/) `>= 3.6.3`
-
-### Installation
-
-#### Manual
+### Manual
 
 1. Download the
    console [aem-groovy-console-all](https://github.com/orbinson/aem-groovy-console/releases/download/18.0.0/aem-groovy-console-all-18.0.0.zip)
```
Comment (bdhoine) [inline README.md:36]: should we keep these versions inline? if so I guess the build should update it

File: README.md [hunk 2]
```diff
@@ -45,7 +39,24 @@ project [aem-groovy-console](https://github.com/CID15/aem-groovy-console)
...
+### Maven profiles
+
+Maven profiles can be used to install the bundles to AEM / Sling
+
+* AEM Author running on localhost:4502
+  * api & bundle: `-P autoInstallBundle`
+  * ui.apps, ui.apps.aem, ui.config, ui.content: `-P autoInstallPackage`
+  * all: `-P autoInstallSinglePackage`
+* AEM Publish running on localhost:4503
+  * api & bundle: `-P autoInstallBundle,publish`
```
Comment (bdhoine) [inline README.md:51]: usually `autoInstallPackagePublish` is used as convention

File: bundle/src/main/groovy/be/orbinson/aem/groovy/console/extension/impl/binding/AemBindingExtensionProvider.groovy [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Override
+    Map<String, BindingVariable> getBindingVariables(ScriptContext scriptContext) {
+        def resourceResolver = scriptContext.resourceResolver
+        def session = resourceResolver.adaptTo(Session)
+
+        def bindingVariables = [
+                pageManager: new BindingVariable(resourceResolver.adaptTo(PageManager), PageManager),
+                pageBuilder: new BindingVariable(new PageBuilder(session), PageBuilder,
+                        "                        https://orbinson.github.io/aem-groovy-console/be/orbinson/aem/groovy/console/builders/PageBuilder.html"),
```
Comment (bdhoine) [inline bundle/src/main/groovy/be/orbinson/aem/groovy/console/extension/impl/binding/AemBindingExtensionProvider.groovy:28]: trim whitespace?

File: bundle/src/main/groovy/be/orbinson/aem/groovy/console/configuration/impl/DefaultConfigurationService.groovy [hunk 4]
```diff
@@ -104,12 +104,18 @@ class DefaultConfigurationService implements ConfigurationService {
...
+            def userManager = resourceResolver.adaptTo(UserManager);
+            if (userManager != null) {
+                def user = resourceResolver.adaptTo(UserManager).getAuthorizable(request.userPrincipal) as User
+                def memberOfGroupIds = user.memberOf()*.ID
+
+                LOG.debug("member of group IDs : {}, allowed group IDs : {}", memberOfGroupIds, groupIds)
+
+                user.admin || (groupIds ? memberOfGroupIds.intersect(groupIds as Iterable) : false)
+            } else {
+                LOG.debug("UserManager not available, probably in a Sling Based application, reverting to is admin check")
```
Comment (bdhoine) [inline bundle/src/main/groovy/be/orbinson/aem/groovy/console/configuration/impl/DefaultConfigurationService.groovy:116]: falling back to

File: bundle/src/main/groovy/be/orbinson/aem/groovy/console/extension/impl/binding/JcrBindingExtensionProvider.groovy [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    @Override
+    Map<String, BindingVariable> getBindingVariables(ScriptContext scriptContext) {
+        def resourceResolver = scriptContext.resourceResolver
+        def session = resourceResolver.adaptTo(Session)
+
+        def bindingVariables = [
+                session    : new BindingVariable(session, Session,
+                        "https://developer.adobe.com/experience-manager/reference-materials/spec/javax.jcr/javadocs/jcr-2.0/javax/jcr/Session.html"),
+                nodeBuilder: new BindingVariable(new NodeBuilder(session), NodeBuilder,
+                        "                        https://orbinson.github.io/aem-groovy-console/be/orbinson/aem/groovy/console/builders/NodeBuilder.html")
```
Comment (bdhoine) [inline bundle/src/main/groovy/be/orbinson/aem/groovy/console/extension/impl/binding/JcrBindingExtensionProvider.groovy:23]: trim?

File: bundle/src/main/groovy/be/orbinson/aem/groovy/console/extension/impl/startimport/SlingStarImportExtensionProvider.groovy [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+class SlingStarImportExtensionProvider implements StarImportExtensionProvider {
+
+    private static final String SLING_JAVADOC_PREFIX = "https://sling.apache.org/apidocs/sling12"
+
+    private static final String JAVADOC_SUFFIX = "package-summary.html"
+
+    private static final Set<StarImport> IMPORTS = ImmutableSet.of(
+            new StarImport("org.apache.sling.api", "$SLING_JAVADOC_PREFIX/org/apache/sling/api/$JAVADOC_SUFFIX"),
+            new StarImport("org.apache.sling.api.resource",
+                    "$SLING_JAVADOC_PREFIX/org/apache/sling/api/resource/$JAVADOC_SUFFIX")
```
Comment (bdhoine) [inline bundle/src/main/groovy/be/orbinson/aem/groovy/console/extension/impl/startimport/SlingStarImportExtensionProvider.groovy:21]: add distribution import?

File: bundle/src/main/groovy/be/orbinson/aem/groovy/console/impl/AEMDetector.groovy [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+import com.day.cq.wcm.api.PageManagerFactory
+import groovy.util.logging.Slf4j
+import org.osgi.service.component.annotations.Component
+import org.osgi.service.component.annotations.Reference
+
+@Component(
+        service = AEMDetector.class,
+        immediate = true
+)
+@Slf4j("LOG")
```
Comment (bdhoine) [inline bundle/src/main/groovy/be/orbinson/aem/groovy/console/impl/AEMDetector.groovy:13]: log not used

File: bundle/src/main/groovy/be/orbinson/aem/groovy/console/servlets/ReplicateScriptServlet.groovy [hunk 8]
```diff
@@ -29,6 +29,9 @@ import static javax.servlet.http.HttpServletResponse.SC_FORBIDDEN
 ])
 @Slf4j("LOG")
 class ReplicateScriptServlet extends AbstractJsonResponseServlet {
+
+    // TODO add check?
```
Comment (bdhoine) [inline bundle/src/main/groovy/be/orbinson/aem/groovy/console/servlets/ReplicateScriptServlet.groovy:33]: add what checks?

File: ui.apps/src/main/content/jcr_root/apps/groovyconsole/clientlibs/css/datatables.css [hunk 9]
```diff
@@ -272,12 +272,6 @@ th {
     border: 1px solid #ddd !important;
   }
 }
-@font-face {
```
Comment (bdhoine) [inline ui.apps/src/main/content/jcr_root/apps/groovyconsole/clientlibs/css/datatables.css:275]: ~~just delete? :) easy fix :D~~

File: ui.apps/src/main/content/jcr_root/apps/groovyconsole/clientlibs/js/console.js [hunk 10]
```diff
@@ -173,7 +180,7 @@ var GroovyConsole = function () {
 
           $('#schedule-job-text').text('Scheduling...');
 
-          $.post(CQ.shared.HTTP.getContextPath() + '/bin/groovyconsole/jobs.json', {
+          $.post(GroovyConsole.getContextPath() + '/bin/groovyconsole/jobs.json', {
```
Comment (bdhoine) [inline ui.apps/src/main/content/jcr_root/apps/groovyconsole/clientlibs/js/console.js:183]: create var for this and interpolate in js string?

File: ui.apps/src/main/content/jcr_root/apps/groovyconsole/components/console/head.html [hunk 11]
```diff
@@ -1,5 +1,16 @@
...
-      data-sly-use.head="/libs/wcm/foundation/components/page/head.js"
-      data-sly-use.headlibRenderer="headlibs.html"/>
-<sly data-sly-call="${headlibRenderer.headlibs @ designPath=head.designPath}" />
-<title>${head.title}</title>
\ No newline at end of file
+<meta http-equiv="content-type" content="text/html; charset=UTF-8"/>
+<style>
+@font-face {
+  font-family: 'Glyphicons Halflings';
+
```
Comment (bdhoine) [inline ui.apps/src/main/content/jcr_root/apps/groovyconsole/components/console/head.html:5]: here they are :D

File: all/pom.xml [hunk 12]
```diff
@@ -83,8 +89,31 @@
     </build>
 
     <profiles>
+        <!-- Default package is built for AEM -->
+        <profile>
+            <id>aem</id>
+            <activation>
+                <activeByDefault>true</activeByDefault>
```
Comment (bdhoine) [inline all/pom.xml:96]: check if build definition needs update because when passing a profile the active by default profiles won't be active ;)

**Review by bdhoine — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (royteeuwen) [reply README.md:36]: sure :D but wont do it in this PR though ^^

File: README.md (hunk 2 above)
Comment (royteeuwen) [reply README.md:51]: I know, but I hate the project archetype convention :( it forces you to duplicate everything. If you rather go for the convention, then I'll duplicate everything twice (once for publish and once for sling)

File: ui.apps/src/main/content/jcr_root/apps/groovyconsole/clientlibs/js/console.js (hunk 10 above)
Comment (royteeuwen) [reply ui.apps/src/main/content/jcr_root/apps/groovyconsole/clientlibs/js/console.js:183]: it's used all over the places, not sure if that will make it easier than this :)

File: bundle/src/main/groovy/be/orbinson/aem/groovy/console/servlets/ReplicateScriptServlet.groovy (hunk 8 above)
Comment (royteeuwen) [reply bundle/src/main/groovy/be/orbinson/aem/groovy/console/servlets/ReplicateScriptServlet.groovy:33]: done

File: bundle/src/main/groovy/be/orbinson/aem/groovy/console/extension/impl/startimport/SlingStarImportExtensionProvider.groovy (hunk 6 above)
Comment (royteeuwen) [reply bundle/src/main/groovy/be/orbinson/aem/groovy/console/extension/impl/startimport/SlingStarImportExtensionProvider.groovy:21]: done

File: README.md (hunk 1 above)
Comment (bdhoine) [reply README.md:36]: acceptable xD

File: README.md (hunk 2 above)
Comment (bdhoine) [reply README.md]: +1

File: ui.apps/src/main/content/jcr_root/apps/groovyconsole/clientlibs/js/console.js (hunk 10 above)
Comment (bdhoine) [reply ui.apps/src/main/content/jcr_root/apps/groovyconsole/clientlibs/js/console.js:183]: if not more readable we should use this indeed

File: .github/workflows/build-on-prem.yml [hunk 13]
```diff
@@ -25,4 +25,4 @@ jobs:
             ${{ runner.os }}-maven-
 
       - name: Build for AEM 6.5
-        run: mvn --batch-mode --update-snapshots clean install
```
Comment (bdhoine) [inline .github/workflows/build-on-prem.yml:28]: rename to build and update README with new badge and delete deleted workflow badge

File: bundle/src/main/groovy/be/orbinson/aem/groovy/console/extension/impl/scriptmetaclass/AemScriptMetaClassExtensionProvider.groovy [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+            delegate.activate = { String path, ReplicationOptions options = null ->
+                replicator.replicate(session, ReplicationActionType.ACTIVATE, path, options)
+            }
+
+            delegate.deactivate = { String path, ReplicationOptions options = null ->
+                replicator.replicate(session, ReplicationActionType.DEACTIVATE, path, options)
+            }
+
+            delegate.invalidate { String path, ReplicationOptions options = null ->
```
Comment (bdhoine) [inline bundle/src/main/groovy/be/orbinson/aem/groovy/console/extension/impl/scriptmetaclass/AemScriptMetaClassExtensionProvider.groovy:45]: added to docs page?

**Review by bdhoine — CHANGES_REQUESTED** [review]

File: bundle/src/main/groovy/be/orbinson/aem/groovy/console/extension/impl/scriptmetaclass/AemScriptMetaClassExtensionProvider.groovy [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+                replicator.replicate(session, ReplicationActionType.DELETE, path, options)
```
Comment (bdhoine) [inline bundle/src/main/groovy/be/orbinson/aem/groovy/console/extension/impl/scriptmetaclass/AemScriptMetaClassExtensionProvider.groovy:46]: why is this cloud only?

**Review by bdhoine — APPROVED** [review]

File: bundle/src/main/groovy/be/orbinson/aem/groovy/console/extension/impl/scriptmetaclass/AemScriptMetaClassExtensionProvider.groovy (hunk 15 above)
Comment (royteeuwen) [reply bundle/src/main/groovy/be/orbinson/aem/groovy/console/extension/impl/scriptmetaclass/AemScriptMetaClassExtensionProvider.groovy:46]: hmm nvm, we used to use the distribution api, and that was cloud only!

**Pull request merged**